### PR TITLE
Add suboption to Hidden (HD) for decreasing visibility based on combo

### PR DESCRIPTION
**Pull request by DavidBeh**
Video:

https://www.youtube.com/watch?v=Hc0kffA_Ewk

This implementation addresses #30871 for osu!std.

When a player's combo breaks, all approach circles on the screen become visible. As the player's combo increases again, the visibility of new approach circles gradually decreases until they disappear completely once the combo reaches the value specified by the player. The mod tracks and uses an internal combo value that is influenced solely by the results of HitCircles and SliderHeads.

This behavior can also be easily adapted for hidden mods in other rulesets. ModHidden handles most of the necessary logic and provides extensions for implementers to adjust key aspects, after which they only need to apply the opacity provided by ModHidden.

### CODE REVIEW

**Comment by bdach** [comment]
I dunno if we want this in the game. Leaning no by default unless there is actual support.

**Comment by DavidBeh** [comment]
Thanks for taking a look.
> Leaning no by default unless there is actual support.

Support from whom? Apart from your input, I’ve only seen neutral or positive reactions in both the GitHub Discussion and the Development Discord Server. Since the option is disabled by default, it would be unintrusive to players.

Additionally, the concept of adjusting a mod's difficulty based on player performance is already well-established in the game. Even some ranked mods, like Blinds, have similar mechanics. (Also, I don’t necessarily want this option to be ranked.)

Personally, I believe this feature offers a great way to ease into playing the hidden mod or practicing beatmaps with it enabled.

Are there specific concerns you have with the concept, the behavior of the implementation, or the code itself?

Also, I am aware of the failing CI tests. I will look into these in the following two weeks.

**Comment by bdach** [comment]
> Are there specific concerns you have with the concept, the behavior of the implementation, or the code itself?

I have only one concern and it's very simple: it's extra complexity and I don't want extra complexity if there isn't a very good reason for it.

What you got in terms of support is 3 upvotes on the discord discussion. I'm not sure what happened on discord because it's an ephemeral communication medium in my mind and I don't 100% keep up with it.

I would want to see a considerably higher number of supporters in order to consider adding this feature.

**Comment by DavidBeh** [comment]
Is [this ](https://osu.ppy.sh/community/forums/topics/2014171?n=1) enough support?

**Comment by bdach** [comment]
Results look 50-50ish. I'm not particularly positively swayed by that result, so I guess not.

**Comment by DavidBeh** [comment]
> Results look 50-50ish. I'm not particularly positively swayed by that result

I only added negative voting options for statistical/transparancy purposes and I don't think that the ratio should serve as an argument for weither the feature should be implemented.

The reason is, that the feature is disabled by default and the only other effect is [a second control](https://live.staticflickr.com/65535/54209258370_cbf3e5c5bf_o.png) for this mods preferences-menu. Nobody can seriously be disrupted by this. So the only reasons for people to vote against are:
* they would not use it, because don't need it/no lazer/no std/etc...
* they only read the title (or other misunderstandings)
* adds too much complexity

Apart from the last one, these are not valid arguments adding this feature. And as for complexity, almost every addition adds complexity. And I think in this case, where it's just in a corner of the codebase and well documented by me, this argument looks weak.

In conclusion, counting all sources there have already been more than 10 people expressing support for this (i admit partially because of me promoting it on various channels), which is the only useful metric we got, so why can't we move forward with this?

**Comment by ominoussage** [comment]
Looks cool but having some circles not hidden shouldn't have any score multiplier bonus and the score bonus should also adjust based on how Hidden the circles are.

It's kinda complicated but if you manage to do that, then this'll make Hidden playable for everybody.

**Comment by DavidBeh** [comment]
I am happy you like the concept, but there are a few problems with your
suggestion. First of all, I intend score to be unranked with a 1x
multiplier when the feature is enabled. Refer to the image I linked in my
last comment. It's handled the same way with non-default flashlight sizes,
so it gives no advantage over other scores.

In my opinion plays with different mods are just not comparable, so it is
hard for me to come up with how to calculate the score multiplier.

There would be some limitations however:
* No score multiplier based on the beatmap; the osu team is absolutely
against this as it increases complexity tremendously.
* (I assume) No score multiplier based on the, because it has a similarly
high complexity.

So it would be up to 3rd parties to rank different plays of this kind (but
I am not sure if osu offers apis for this)

There is however a toggle in lazers beatmap selection page for only showing
plays with the selected currently mods and their preferences in
leaderboards. It's only available to supporters though. It would make sense
to offer only allow a few options (like 0, 5, 10, 15) for the combo at
which hit objects are hidden, so more scores can be shown when filtering.

In conclusion what you are suggesting is cool, but for now out of scope for
this PR. For now I intend it to be played for fun and for easing into
starting to play with hd or easing into starting to play new beatmap with
hd.

ominoussage ***@***.***> schrieb am Di., 24. Dez. 2024, 03:18:

**Comment by ominoussage** [comment]
Sorry, it wasn't obvious at first glance that the link from your last comment was an image that addresses my concerns.

As long as it's unranked, everything's good.

File: osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs [hunk 1]
```diff
@@ -68,18 +87,17 @@ private void applyHiddenState(DrawableHitObject drawableObject, bool increaseVis
...
-                {
-                    using (circle.BeginAbsoluteSequence(hitObject.StartTime - hitObject.TimePreempt))
-                        circle.ApproachCircle.Hide();
-                }
-                else if (drawableObject is DrawableSpinner spinner)
-                {
-                    spinner.Body.OnSkinChanged += () => hideSpinnerApproachCircle(spinner);
-                    hideSpinnerApproachCircle(spinner);
-                }
+                float alpha = increaseVisibility ? 1 : GetAndUpdateDrawableHitObjectComboAlpha(circle2);
```
Comment (DavidBeh) [inline osu.Game.Rulesets.Osu/Mods/OsuModHidden.cs:92]: always returns 0 when feature is disabled
